utils/build-script

```diff
@@ -63,7 +62,718 @@ def call_without_sleeping(command, dry_run=False):
+                    "--install-symroot is required when specifying "
+                    "--symbols-package.")
+
+        if args.android:
+            if args.android_ndk is None or \
+                    args.android_api_level is None or \
+                    args.android_icu_uc is None or \
+                    args.android_icu_uc_include is None or \
+                    args.android_icu_i18n is None or \
+                    args.android_icu_i18n_include is None:
+                diagnostics.fatal(
+                    "when building for Android, --android-ndk, "
+                    "--android-ndk-version, --android-icu-uc, "
+                    "--android-icu-uc-include, --android-icu-i18n, "
+                    "and --android-icu-i18n-include must be specified")
+
+    @staticmethod
+    def apply_default_arguments(toolchain, args):
+        """Preprocess an argument set to apply default behaviors."""
+
+        # Build cmark if any cmark-related options were specified.
+        if (args.cmark_build_variant is not None):
+            args.build_cmark = True
+
+        # Build LLDB if any LLDB-related options were specified.
+        if args.lldb_build_variant is not None or \
+           args.lldb_assertions is not None:
+            args.build_lldb = True
+
+        # Set the default build variant.
+        if args.build_variant is None:
+            args.build_variant = "Debug"
+
+        # Propagate the default build variant.
+        if args.cmark_build_variant is None:
+            args.cmark_build_variant = args.build_variant
+
+        if args.llvm_build_variant is None:
+            args.llvm_build_variant = args.build_variant
+
+        if args.swift_build_variant is None:
+            args.swift_build_variant = args.build_variant
+
+        if args.swift_stdlib_build_variant is None:
+            args.swift_stdlib_build_variant = args.build_variant
+
+        if args.lldb_build_variant is None:
+            args.lldb_build_variant = args.build_variant
+
+        if args.foundation_build_variant is None:
+            args.foundation_build_variant = args.build_variant
+
+        if args.libdispatch_build_variant is None:
+            args.libdispatch_build_variant = args.build_variant
+
+        # Assertions are enabled by default.
+        if args.assertions is None:
+            args.assertions = True
+
+        # Propagate the default assertions setting.
+        if args.cmark_assertions is None:
+            args.cmark_assertions = args.assertions
+
+        if args.llvm_assertions is None:
+            args.llvm_assertions = args.assertions
+
+        if args.swift_assertions is None:
+            args.swift_assertions = args.assertions
+
+        if args.swift_stdlib_assertions is None:
+            args.swift_stdlib_assertions = args.assertions
+
+        # Set the default CMake generator.
+        if args.cmake_generator is None:
+            args.cmake_generator = "Ninja"
+
+        ninja_required = (
```

Comment:
Everything above this line doesn't depend on the actual arguments that were passed in. I would prefer to have these as the initialised state of a `BuildScriptInvocation` object and to move the more complex argument-parsing logic below with the other argument-parsing logic.

Otherwise I'm afraid we'll end up with really complex webs of 'defaults' which get hard to untangle. Defaults mean the plain, initialised state. Anything that depends on the arguments that were passed in is parsing logic, not default-logic IMO.